libgen/utils.go

```diff
@@ -0,0 +1,649 @@
+	for _, cap := range defaultCaps {
+		addedCapsMap[cap] = true
+	}
+
+	addCapList := make([]string, len(defaultCaps))
+	copy(addCapList, defaultCaps)
+	for _, c := range addCaps {
+		if !capMappings[c] {
+			return fmt.Errorf("Invalid value passed for adding capability")
+		}
+		cp := fmt.Sprintf("CAP_%s", c)
+		if !addedCapsMap[cp] {
+			addCapList = append(addCapList, cp)
+			addedCapsMap[cp] = true
+		}
+	}
+	dropCapsMap := make(map[string]bool)
+	for _, c := range dropCaps {
+		if !capMappings[c] {
+			return fmt.Errorf("Invalid value passed for dropping capability")
+		}
+		cp := fmt.Sprintf("CAP_%s", c)
+		dropCapsMap[cp] = true
+	}
+
+	for _, c := range addCapList {
+		if !dropCapsMap[c] {
+			finalCapList = append(finalCapList, c)
+		}
+	}
+	spec.Process.Capabilities = finalCapList
+	return nil
+}
+
+func mapStrToNamespace(ns string, path string) rspec.Namespace {
+	switch ns {
+	case "network":
+		return rspec.Namespace{Type: rspec.NetworkNamespace, Path: path}
+	case "pid":
+		return rspec.Namespace{Type: rspec.PIDNamespace, Path: path}
+	case "mount":
+		return rspec.Namespace{Type: rspec.MountNamespace, Path: path}
+	case "ipc":
+		return rspec.Namespace{Type: rspec.IPCNamespace, Path: path}
+	case "uts":
+		return rspec.Namespace{Type: rspec.UTSNamespace, Path: path}
+	case "user":
+		return rspec.Namespace{Type: rspec.UserNamespace, Path: path}
+	case "cgroup":
+		return rspec.Namespace{Type: rspec.CgroupNamespace, Path: path}
+	default:
+		logrus.Fatalf("Should not reach here!")
+	}
+	return rspec.Namespace{}
+}
+
+func checkNs(nsMaps map[string]string, nsName string) bool {
+	if _, ok := nsMaps[nsName]; !ok {
+		return false
+	}
+	return true
+}
+
+// SetupNamespaces adds the namespace settings into spec.
+func SetupNamespaces(spec *rspec.Spec, needsNewUser bool, nsMaps map[string]string) {
+
+	for _, nsName := range Namespaces {
+		if !checkNs(nsMaps, nsName) && !(needsNewUser && nsName == "user") {
+			continue
+		}
+		nsPath := nsMaps[nsName]
+		if nsPath == "host" {
+			ns := mapStrToNamespace(nsName, "")
+			removeNamespace(&spec.Linux.Namespaces, ns.Type)
+			continue
+		}
+		ns := mapStrToNamespace(nsName, nsPath)
+		replaceOrAppendNamespace(&spec.Linux.Namespaces, ns)
+	}
+}
+
+func replaceOrAppendNamespace(namespaces *[]rspec.Namespace, namespace rspec.Namespace) {
+	for i, ns := range *namespaces {
+		if ns.Type == namespace.Type {
+			(*namespaces)[i] = namespace
+			return
+		}
+	}
+	new := append(*namespaces, namespace)
+	*namespaces = new
+}
+
+func removeNamespace(namespaces *[]rspec.Namespace, namespaceType rspec.NamespaceType) {
+	for i, ns := range *namespaces {
+		if ns.Type == namespaceType {
+			*namespaces = append((*namespaces)[:i], (*namespaces)[i+1:]...)
+			return
+		}
+	}
+}
+
+// StrPtr returns the pointer pointing to the string s.
+func StrPtr(s string) *string { return &s }
+
+// GetDefaultTemplate returns the default spec.
+func GetDefaultTemplate() *rspec.Spec {
```

Comment:
I'd put this in `default.go`.  And does Go have a deep copy?  If so, I'd consider making this a variable.  That way it's easy for folks to swap in an alternative default template or mutate the original if they want, and they can make a deep copy if they don't want to mutate the default template.